### PR TITLE
Active search result sticks to bottom and top of page

### LINKED ISSUE #276: When arrowing down through results, the active result should be "stuck" to the bottom
canjs_search_highlight_keyboard

### PR DESCRIPTION
**Pull request by imaustink**
search-result-stuck

### CODE REVIEW

**Comment by chasenlehara** [comment]
Going down is looking good!

I think we should make it behave the same way going up: when the user hits the up arrow when the last item is selected, what’s visible in the scroll area should stay the same while the selection moves up. See the gif below for the unwanted behavior:
going up

**Comment by imaustink** [comment]
@chasenlehara this PR has been updated to address your last comment.
(gray is too hard to see so I made it red for the gif)
search-result-stuck

File: static/search.js [hunk 1]
```diff
@@ -674,13 +674,40 @@ var Search = Control.extend({
...
-		this.$resultsContainer.scrollTop(this.$activeResult.position().top - activeResultOffset);
+		// Get position top of current active element
+		var activeResultPosTop = parseInt(this.$activeResult.position().top);
+
+		var isMovingDown = lastResultPosTop < this.$activeResult.position().top;
+		
+		if(isMovingDown){
+			var resultsContainerHeight = this.$resultsContainer.height();
+
+			var isBellow = activeResultPosTop > resultsContainerHeight;
```
Comment (bgando) [inline static/search.js]: isBelow?

File: static/search.js [hunk 2]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
 	// function activateResult
 	// sets property and adds class to active result
 	activateResult: function($result){
+		// Get position top of last active element
+		var lastResultPosTop = parseInt(this.$activeResult && this.$activeResult.position().top || 0);
```
Comment (bgando) [inline static/search.js]: props on IDing this bug ahead of time!

File: static/search.js [hunk 3]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
...
+		var activeResultPosTop = parseInt(this.$activeResult.position().top);
+		var resultsContainerHeight = this.$resultsContainer.height();
 
-		this.$resultsContainer.scrollTop(this.$activeResult.position().top - activeResultOffset);
+		var isMovingDown = lastResultPosTop < this.$activeResult.position().top;
+		var isBellow = activeResultPosTop > resultsContainerHeight;
+		var isAbove = activeResultPosTop < 0;
+
+		if(isMovingDown){
+			if(isBellow){
```
Comment (bgando) [inline static/search.js]: isMovingDown && isBelow could be on one line

File: static/search.js [hunk 4]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
...
+				this.resetScrollToBottom(activeResultPosTop, resultsContainerHeight);
+			}
+		}else{
+			if(isAbove){
```
Comment (bgando) [inline static/search.js]: could use an else if(isAbove)

File: static/search.js [hunk 5]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
...
+		var resultsContainerHeight = this.$resultsContainer.height();
 
-		this.$resultsContainer.scrollTop(this.$activeResult.position().top - activeResultOffset);
+		var isMovingDown = lastResultPosTop < this.$activeResult.position().top;
+		var isBellow = activeResultPosTop > resultsContainerHeight;
+		var isAbove = activeResultPosTop < 0;
+
+		if(isMovingDown){
+			if(isBellow){
+				this.resetScrollToBottom(activeResultPosTop, resultsContainerHeight);
```
Comment (bgando) [inline static/search.js]: Much more readable! Awesome 💃

File: static/search.js [hunk 6]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
...
 		this.$activeResult = $result;
 		this.$activeResult.addClass(this.options.keyboardActiveClass);
 
-		var activeResultOffset = this.getActiveResultOffset();
+		// Get position top of current active element
+		var activeResultPosTop = parseInt(this.$activeResult.position().top);
+		var resultsContainerHeight = this.$resultsContainer.height();
 
-		this.$resultsContainer.scrollTop(this.$activeResult.position().top - activeResultOffset);
+		var isMovingDown = lastResultPosTop < this.$activeResult.position().top;
```
Comment (bgando) [inline static/search.js]: Add a comment here to clarify that you are talking about the arrow key direction

File: static/search.js [hunk 7]
```diff
@@ -674,13 +674,45 @@ var Search = Control.extend({
...
+		var isBellow = activeResultPosTop > resultsContainerHeight;
```
Comment (bgando) [inline static/search.js]: isBelow

**Comment by imaustink** [comment]
@bgando I agree with your reasoning behind splitting up `activateResult`, the only problem is that part of the math for scrolling involves the position top of the previous active element.

**Comment by imaustink** [comment]
@chasenlehara do you want it to stop when it reached the bottom or top? Currently, it restarts from the other end.

**Comment by chasenlehara** [comment]
I personally don’t like that it goes back to the top when you reach the bottom, but I’m willing to see how people react to it in practice.

File: static/search.js (hunk 2 above)
Comment (chasenlehara) [inline static/search.js]: Make sure to always pass `10` as the second argument to `parseInt` (you can search for reasons why).

**Pull request merged**